File: SRC/framework/JSBundle.m
//
//  JSBundle.m
//  test
//
//  Created by yineng on 2018/8/9.
//  Copyright © 2018年 kupay. All rights reserved.
//

#import "JSBundle.h"
#import "WebViewController.h"

@interface JSBundle ()

@end

static int identity = 0;

// <int, id>
static NSMutableDictionary *dictionary = nil;

@implementation JSBundle

+ (void)callJS:(NSNumber *)listenerID code:(CallJSType)code params:(NSArray *)params {
    
    NSLog(@"CallJS");
    
    NSString *s = [NSString stringWithFormat:@"window['handle_Native_Message'](%@, %d,", listenerID, code];
    NSMutableArray *array = [NSMutableArray arrayWithObjects:s, nil];
  
    NSInteger count = [params count];
    for (int i = 0 ; i < count; ++i) {
        
        id obj = [params objectAtIndex:i];
    
        NSString *str = nil;
        if ([obj isKindOfClass:[NSString class]]) {
            str = [NSString stringWithFormat:@"'%@'", (NSString *)obj];
        } else {
            str = [NSString stringWithFormat:@"%@", obj];
        }
        
        if (i + 1 < count) {
            str =  [str stringByAppendingString:@", "];
        }
        
        [array addObject:str];
    }
    
    [array addObject:@")"];
    
    NSString *fullCode = [array componentsJoinedByString:@""];
    WKWebView *webview = [WebViewController getWebView];
    
    // NSLog(@"callJS fullCode = %@", fullCode);
    [webview evaluateJavaScript:fullCode completionHandler:^(id object,NSError *error) {
        if(error != nil) {
            NSLog(@"item = %@, error = %@", object, error);
        }
    }];
}

+ (void)callJSError:(NSString *)arrayClassName funcName:(NSString *)funcName msg:(NSString *)msg {
    
    NSLog(@"CallJSError");
    
    WKWebView *webview = [WebViewController getWebView];
    NSString *code = [NSString stringWithFormat:@"window['handle_Native_ThrowError']('%@','%@','%@')", arrayClassName, funcName, msg];
    // NSLog(@"callJSError code = %@", code);
     [webview evaluateJavaScript:code completionHandler:^(id object,NSError *error) {
        if(error != nil) {
            NSLog(@"item= %@, error= s%@", object, error);
        }
    }];
}

// 收到JS的消息，做分发
+ (void)sendMessage:(NSArray *)params {
    
    NSLog(@"sendMessage: %@", params);
    
    if (dictionary == nil) {
        dictionary = [NSMutableDictionary dictionaryWithCapacity:10];
    }
    
    NSString *className = params[0];
    NSString *funcName = params[1];
    NSNumber *nativeID = params[2];
    
    params = [params subarrayWithRange:NSMakeRange(3, [params count] - 3)];
    
    Class clazz = NSClassFromString(className);
    if (clazz == nil) {
        NSLog(@"JSBundle Error = className");
        return;
    }

    if (nativeID.intValue == 0) {
        if ([funcName isEqualToString:@"init"]) {
            [JSBundle initImpl:clazz params:params];
        } else {
            // 类方法
            NSString *s = [funcName stringByAppendingString:@":"];
            SEL sel = NSSelectorFromString(s);
            if ([clazz respondsToSelector:sel]) {
                [clazz performSelector:sel withObject:params];
            } else {
                [JSBundle callJSError:NSStringFromClass(clazz) funcName:funcName msg:@"class methed funcName not found"];
            }
        }
    } else {
        // close方法
        if ([funcName isEqualToString:@"close"]) {
            [dictionary removeObjectForKey:nativeID];
            return;
        }
        
        // 其他实例方法
        NSObject *object = [dictionary objectForKey:nativeID];
        if (object == nil) {
            NSLog(@"JSBundle Error = object not found");
            return;
        }

        NSString *s = [funcName stringByAppendingString:@":"];
        SEL sel = NSSelectorFromString(s);
        if ([object respondsToSelector:sel]) {
            [object performSelector:sel withObject:params];
        } else {
            [JSBundle callJSError:NSStringFromClass(clazz) funcName:funcName msg:@"object method funcName not found"];
        }
    }
}

// 初始化对象并添加到字典中
+ (void)initImpl: (Class)clazz params: (NSArray*)array {
    id obj = [[clazz alloc] init];
    if (obj == nil) {
        [JSBundle callJSError:NSStringFromClass(clazz) funcName:@"init" msg:@"init failed"];
    }
    
    NSNumber *listenerID = [array objectAtIndex:0];
    if (listenerID == nil) {
        [JSBundle callJSError:NSStringFromClass(clazz) funcName:@"init" msg:@"listenerID not found"];
    }
    
    identity += 1;
    [dictionary setObject:obj forKey:[NSNumber numberWithInt:identity]];
    [JSBundle callJS:listenerID code:0 params:[NSArray arrayWithObjects:[NSNumber numberWithInt:identity],nil]];
}

@end

File: test/JSBundle.swift
//
//  JSBundle.swift
//  tbsWebView
//
//  Created by Apple on 17/9/1.
//  Copyright © 2017年 Ganchu. All rights reserved.
//

import UIKit

//MARK: callJS的状态
enum callJSType:Int{
    case Success = 0
    case Fail = 1
    case Callback = 100
}

class JSBundle: NSObject {
    
    var nameArray:[String:Any] = [:]
    
    static var id = 0;
    
    static var dictionary:[Int:Any] = [Int:Any]()
    
    //MARK: 初始化对象并添加到字典中
    func initImpl(_ classType: NSObject.Type,_ array:Array<Any> )  {
        let obj = classType.init()
        let listenerId = array[0] as! Int
        JSBundle.id += 1
        JSBundle.dictionary.updateValue(obj, forKey: JSBundle.id)
        JSBundle.callJS(listenerId,callJSType.Success,JSBundle.id)
    }
    
    //MARK: 收到JS的消息，做分发
    func sendMessage(_ json:Any) {
        var array = json as! Array<Any>
        let arrayClassName = array[0] as! String
        let funcName = array[1] as! String
        let nativeID = array[2] as! Int
        array.removeFirst(3)
        
        guard let stringFromClassName = nameArray["\(arrayClassName)"]  else {
            print("JSBunldleError = class名字错误")
            return
        }
        
        let className = NSStringFromClass(stringFromClassName as! AnyClass)
        guard  let classType = NSClassFromString(className) as? NSObject.Type else {
            print("JSBunldleError = class实例化失败")
            return
        }
        
        if nativeID == 0 {
            if funcName == "init" {
                initImpl(classType,array)
            }else {
                classType.perform(NSSelectorFromString("\(funcName):"), with: array)
            }
        }else {
            
            if funcName == "close" {
                JSBundle.dictionary.removeValue(forKey: nativeID)
                return
            }
            // 寻找hash表对象
            guard let cls = JSBundle.dictionary[nativeID] as? NSObject else {
                print("JSBunldleError = cls取值为空")
                return
            }
            if cls.responds(to: NSSelectorFromString("\(funcName):")) {
                cls.perform(NSSelectorFromString("\(funcName):"), with: array)
            }else {
                JSBundle.callJSError(arrayClassName, funcName, "")
            }
        }
    }
    
    // MARK: 封装callJS方法。
    class func callJS(_ id: Any,_ code: callJSType, _ params:Any...)  {
        var sinfo = "window['handle_Native_Message'](\(id),\(code.rawValue),"
        var p = [Any]()
        if let y = params[0] as? Array<Any> {
            p = y
        }else {
            p = params
        }
        for i in p {
            if let j = i as? String {
                sinfo.append("'\(j)',")
            } else {
                sinfo.append("\(i),")
            }
        }
        sinfo.remove(at: sinfo.index(before: sinfo.endIndex))
        sinfo.append(")")

        ViewController.webView.evaluateJavaScript(sinfo) { (item, error) in
            if error != nil {
                print(error as Any)
            }
        }
    }
    // MARK: 封装向js抛出错误的方法
    class func callJSError(_ arrayClassName:String,_ funcName:String,_ msg:String) {
        print("开始抛出error")
        ViewController.webView.evaluateJavaScript("window['handle_Native_ThrowError']('\(arrayClassName)','\(funcName)','ios没有这个方法msg=\(msg)')") { (item, error) in
            if error != nil {
                print("item=\(item),error=\(error)")
            }
        }
    }
}
